src/auth/mysql.rs

```diff
@@ -0,0 +1,362 @@
+        let database_uri: String = String::from(
+            format!("mysql://{}:{}@{}:{}/{}", user, pass, host, port, database)
+        );
+        let authenticator = MySqlAuthenticator{database_uri};
+        match authenticator.test_connection() {
+            Ok(_) => Ok(authenticator),
+            Err(e) => Err(Error::GenericError(e.to_string())),
+        }
+    }
+
+    fn test_connection(&self) -> Result<&'static str, Error> {
+        let _ = self.connect()?;
+        Ok("Connection successful")
+    }
+
+    /// Connects to MySql Server
+    fn connect(&self) -> Result<MysqlConnection, Error> {
+        debug_!("Connecting to Mysql server");
+        let connection = MysqlConnection::establish(&self.database_uri)?;
+        Ok(connection)
+    }
+
+    /// Search for the specified user entry
+    fn search(
+        &self,
+        connection: &MysqlConnection,
+        search_user: &str
+    ) -> Result<Vec<User>, Error> {
+        use super::schema::users::dsl::*;
+        let results = users.filter(username.eq(search_user))
+            .load::<User>(connection)
+            .map_err(|e| Error::GenericError(
+                    format!("Database query failed: {}", e)
+            ))?;
+        Ok(results)
+    }
+
+    /// Hash a password with the salt. See struct level documentation for the algorithm used.
+    // TODO: Write an "example" tool to salt easily
+    pub fn hash_password(password: &str, salt: &[u8]) -> Result<String, Error> {
+        Ok(hex_dump(
+            Self::hash_password_digest(password, salt)?.as_ref(),
+        ))
+    }
+
+    /// Verify that some user with the provided password exists in the CSV database, and the password is correct.
+    /// Returns the payload to be included in a refresh token if successful
+    pub fn verify(
+        &self,
+        username: &str,
+        password: &str,
+        include_refresh_payload: bool,
+    ) -> Result<AuthenticationResult, Error> {
+        let user = {
+            let connection = self.connect()?;
+            let mut user = self.search(&connection, username).map_err(|_e| {
+                super::Error::AuthenticationFailure
+            })?;
+            if user.len() != 1 {
+                Err(super::Error::AuthenticationFailure)?;
+            }
+
+            user.pop().unwrap() // safe to unwrap
+        };
+        if username != user.username {
+            return Err(Error::Auth(super::Error::AuthenticationFailure));
+        }
+        let hash = test::from_hex(&user.pw_hash)?;
+        let salt = test::from_hex(&user.salt)?;
+
+        let actual_password_digest = Self::hash_password_digest(password, &salt)?;
+        if !verify_slices_are_equal(actual_password_digest.as_ref(), &*hash).is_ok() {
+            Err(Error::Auth(super::Error::AuthenticationFailure))
+        } else {
+            let refresh_payload = if include_refresh_payload {
```

Comment:
I think the `refresh_payload` in this case should be a `JsonValue` serialized version of the `User` struct you have defined above instead of the username and password. You can use [this function](https://docs.serde.rs/serde_json/fn.to_value.html) to do the serialization.

Less risky than storing the user's password.